File: 프로그래머스/2/12978. 배달/배달.js
function solution(N, road, K) {
    const infos = Array.from({ length: N + 1 }, () => []);
    const accum = Array(N + 1).fill(2e9);

    road.forEach((info) => {
        let [a, b, c] = info;
        
        infos[a].push({ to: b, time: c }); 
        infos[b].push({ to: a, time: c }); 
    });

    const pq = [{ to: 1, time: 0 }];
    accum[1] = 0;

    while(pq.length !== 0) {
        let {to, time} = pq.pop();

        infos[to].forEach(next => {
            if(accum[next.to] > accum[to] + next.time) {
                accum[next.to] = accum[to] + next.time;
                pq.push(next);
            }
        })
    }

    return accum.filter((item) => {
        return item <= K
    }).length;
}

File: 프로그래머스/2/12978. 배달/배달.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int answer = 1;
vector<vector<pair<int, int>>> info;
vector<int> accumTime;

void BFS(int K) {
    queue<pair<int, int>> q;
   
    q.push({1, 0}); // 1번 마을, 시간 0
    
    while(!q.empty()) {
        int curNode = q.front().first;
        int curTime = q.front().second;
        q.pop();
        
        if(accumTime[curNode] != 0 && accumTime[curNode] < curTime)
            continue;
        if(curTime > K)
            continue;
        
        accumTime[curNode] = curTime;
        
        for(auto next : info[curNode]) {
            if(next.first != 1)
                q.push({next.first, curTime + next.second}); // 다음 노드, 걸린 시간
        }
    }
}

int solution(int N, vector<vector<int>> road, int K) {
    info.resize(N + 1, vector<pair<int, int>>());
    accumTime.resize(N + 1, 0);
    
    for(auto r: road) {
       info[r[0]].push_back(make_pair(r[1], r[2]));
       info[r[1]].push_back(make_pair(r[0], r[2])); 
    }
    
    BFS(K);
    
    for(auto cnt : accumTime) {
        if(cnt)
            answer++;
    }
    
    return answer;
}